src/ScottPlot5/ScottPlot5 Tests/Unit Tests/UnitTests/IDataSourceTests.cs

```diff
@@ -0,0 +1,393 @@
+    {
+        Plot plot = new Plot();
+        var dataSource = ignoreOffsetsAndScaling
+            ? new SignalXYSourceDoubleArray(Xs, Ys)
+            : new SignalXYSourceDoubleArray(Xs, Ys)
+            {
+                XOffset = XOffSet,
+                YOffset = YOffSet,
+                XScale = XScaling,
+                YScale = YScaling
+            };
+        var signal = plot.Add.SignalXY(dataSource);
+        plot.RenderInMemory(PlotWidth, PlotHeight);
+        var mouseLoc = ignoreOffsetsAndScaling ? MouseLocation : DataSourceUtilities.ScaleCoordinate(MouseLocation, XScaling, XOffSet, YScaling, YOffSet);
+
+        var nearest = DataSourceUtilities.GetNearestXFast(dataSource, mouseLoc, plot.LastRender, 100);
+        Assert.That(nearest.Index, Is.EqualTo(IndexToCheck));
+        nearest.X.Should().Be(mouseLoc.X);
+        nearest.Y.Should().BeApproximately(mouseLoc.Y, .001);
+
+        //Native function should also be using this method
+        nearest = signal.GetNearestX(mouseLoc, plot.LastRender, 100);
+        Assert.That(nearest.Index, Is.EqualTo(IndexToCheck));
+        nearest.X.Should().Be(mouseLoc.X);
+        nearest.Y.Should().BeApproximately(mouseLoc.Y, .001);
+    }
+
+    #endregion
+
+    [TestCase(false)]
+    [TestCase(true)]
+    [Test]
+    public void Test_CoordinateDataSource(bool scaling)
+    {
+        // Test IDataSource
+        var ds = new CoordinateDataSource(Cs);
+        if (scaling)
+        {
+            ds.YOffset = YOffSet; 
+            ds.XOffset = XOffSet; 
+            ds.XScale = XScaling; 
+            ds.YScale = YScaling;
+        }
+        Test_IDataSource(ds, scaling ? DataSourceUtilities.ScaleCoordinate(MouseLocation, XScaling, XOffSet, YScaling, YOffSet) : MouseLocation);
+    }
+
+    [TestCase(true)][TestCase(false)][Test] public void Test_SignalSourceDouble(bool scaling) => TestSignal(new SignalSourceDouble(Ys, Period), scaling);
+    [TestCase(true)][TestCase(false)][Test] public void Test_SignalSourceGenericArray(bool scaling) => TestSignal(new SignalSourceGenericArray<float>(GenericYs, Period), scaling);
+    [TestCase(true)][TestCase(false)][Test] public void Test_SignalSourceGenericList(bool scaling) => TestSignal(new SignalSourceGenericList<float>(GenericYs, Period), scaling);
+    [TestCase(true)][TestCase(false)][Test] public void Test_FastSignalSourceDouble(bool scaling) => TestSignal(new FastSignalSourceDouble(Ys, Period), scaling);
+
+    private static void TestSignal<T>(T signalSource, bool scaling) where T : ISignalSource, IDataSource
+    {
+        var mouse = MouseLocation;
+        if (scaling)
+        {
+            signalSource.XOffset = XOffSet;
+            signalSource.YOffset = YOffSet;
+            signalSource.YScale = YScaling;
+            mouse = DataSourceUtilities.ScaleCoordinate(MouseLocation, 1, XOffSet, YScaling, YOffSet);
+        }
+        Console.Write($" - Testing {signalSource.GetType().Name} as {nameof(IDataSource)} ");
```

Comment:
It's not clear what all this console output is for. If you want to use it for debugging, there is Debug.WriteLine() channel. But in general I would just exclude it. We need to know that the test passes or fails, and the test system displays it itself.

Reply:
I'll swap these over to debug. When using console in a test it will show in the test log, and because this test cast a wide net with multiple stages, I had it in place for debugging which stage failed more easily.